save-plugins/warn-plugin/src/commonMain/kotlin/org/cqfn/save/plugin/warn/WarnPlugin.kt

```diff
@@ -0,0 +1,126 @@
+        return sequence {
+            discoverTestFiles(warnPluginConfig.testResources).forEach { testFile ->
+                yield(handleTestFile(testFile, warnPluginConfig, saveProperties))
+            }
+        }
+    }
+
+    /**
+     * Discover test resources for warn-plugin among [resources]
+     *
+     * @param resources a collection of files
+     */
+    internal fun discoverTestFiles(resources: List<Path>) = resources
+        .filter { it.name.contains("Test.") }
+
+    private fun handleTestFile(
+        path: Path,
+        warnPluginConfig: WarnPluginConfig,
+        saveProperties: SaveProperties): TestResult {
+        val expectedWarnings = fs.readLines(path)
+            .mapNotNull {
+                with(warnPluginConfig) {
+                    it.extractWarning(
+                        warningsInputPattern,
+                        columnCaptureGroup,
+                        lineCaptureGroup,
+                        messageCaptureGroup
+                    )
+                }
+            }
+            .groupBy {
+                if (it.line != null && it.column != null) {
+                    it.line to it.column
+                } else {
+                    null
+                }
+            }
+            .mapValues { it.value.sortedBy { it.message } }
+        // todo: create a temp file with technical comments removed and feed it to the tool
+        val executionResult = pb.exec(warnPluginConfig.execCmd.split(" "), null)
+        val actualWarningsMap = executionResult.stdout.mapNotNull {
+            with(warnPluginConfig) {
+                it.extractWarning(warningsOutputPattern, columnCaptureGroup, lineCaptureGroup, messageCaptureGroup)
+            }
+        }
+            .groupBy { if (it.line != null && it.column != null) it.line to it.column else null }
+            .mapValues { (_, warning) -> warning.sortedBy { it.message } }
+        return TestResult(
+            listOf(path),
+            checkResults(expectedWarnings, actualWarningsMap),
+            DebugInfo(executionResult.stdout.joinToString("\n"), executionResult.stderr.joinToString("\n"), null)
+        )
+    }
+
+    @Suppress("TYPE_ALIAS")
+    private fun checkResults(expectedWarningsMap: Map<LineColumn?, List<Warning>>,
+                             actualWarningsMap: Map<LineColumn?, List<Warning>>): TestStatus {
+        checkCollectionsDiffer(expectedWarningsMap, actualWarningsMap)?.let { message ->
+            return Fail(message)
+        }
+        checkWarningsDiffer(expectedWarningsMap, actualWarningsMap)?.let { message ->
+            return Fail(message)
+        }
+        return Pass
+    }
+
+    @Suppress("TYPE_ALIAS")
+    private fun checkCollectionsDiffer(expectedWarningsMap: Map<LineColumn?, List<Warning>>,
+                                       actualWarningsMap: Map<LineColumn?, List<Warning>>): String? {
+        val missingWarnings = expectedWarningsMap.filterValues { it !in actualWarningsMap.values }
```

Comment:
you can map to values here, so in your string you won't need to write ${${missingWarnings.values}}

Reply:
fixed